src/hocon_schema.erl

```diff
@@ -0,0 +1,128 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2021 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(hocon_schema).
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-export([map/2]).
+
+map(Schema, RichMap) ->
+    Keys = apply(Schema, keys, []),
+    do_map(Keys, RichMap, []).
+
+do_map([], _RichMap, Acc) ->
+    Acc;
+do_map([{Key, SchemaFun} | More], RichMap, Acc) ->
+    Value = deep_get(Key, RichMap, value),
+    Value0 = apply_converter(SchemaFun, Value),
+    Validators = add_default_validator(SchemaFun(validator), SchemaFun(type)),
+    Mapping = string:tokens(SchemaFun(mapping), "."),
+    case {Value0, SchemaFun(default)} of
+        {undefined, undefined} ->
+            do_map(More, RichMap, Acc);
+        {undefined, Default} ->
+            do_map(More, RichMap, [{Mapping, Default} | Acc]);
+        {Value0, _} ->
+            Value1 = validate(Value0, Validators),
```

Comment:
handle error(s) returned from `validate/2`
maybe accumulate errors in a list (with both original value and converted values)